libethash-cuda/CUDAMiner.cpp

```diff
@@ -104,21 +106,31 @@ bool CUDAMiner::initEpoch_internal()
             CUDA_SAFE_CALL(cudaDeviceSetCacheConfig(cudaFuncCachePreferL1));
 
             // Check whether the current device has sufficient memory every time we recreate the dag
-            if (m_deviceDescriptor.totalMemory < RequiredMemory)
+            if (m_deviceDescriptor.totalMemory < RequiredTotalMemory)
             {
-                cudalog << "Epoch " << m_epochContext.epochNumber << " requires "
-                        << dev::getFormattedMemory((double)RequiredMemory) << " memory.";
-                cudalog << "This device hasn't available. Mining suspended ...";
-                pause(MinerPauseEnum::PauseDueToInsufficientMemory);
-                return true;  // This will prevent to exit the thread and
-                              // Eventually resume mining when changing coin or epoch (NiceHash)
+                if (m_deviceDescriptor.totalMemory < RequiredDagMemory)
+                {
+                    cudalog << "Epoch " << m_epochContext.epochNumber << " requires "
+                            << dev::getFormattedMemory((double)RequiredDagMemory) << " memory.";
+                    cudalog << "This device hasn't enough memory available. Mining suspended ...";
+                    pause(MinerPauseEnum::PauseDueToInsufficientMemory);
+                    return true;  // This will prevent to exit the thread and
+                                  // Eventually resume mining when changing coin or epoch (NiceHash)
+                }
+                else
+                    lightOnHost = true;
             }
 
-            cudalog << "Generating DAG + Light : "
-                    << dev::getFormattedMemory((double)RequiredMemory);
+            cudalog << "Generating DAG + Light(on " << (lightOnHost ? "host" : "GPU")
+                    << ") : " << dev::getFormattedMemory((double)RequiredTotalMemory);
 
             // create buffer for cache
-            CUDA_SAFE_CALL(cudaMalloc(reinterpret_cast<void**>(&light), m_epochContext.lightSize));
+            if (lightOnHost)
+                CUDA_SAFE_CALL(cudaHostAlloc(reinterpret_cast<void**>(&light),
```

Comment:
Not sure I read it correctly but here you're allocating light for each GPU thread.
Doesn't it clobber host RAM ?

Reply:
I suppose you could use the same host copy of the light cache for all GPUs, but I might be complicated to synchronize all GPUs around epoch changes where you need to regenrate the light cache and dag.

Reply:
I agree sync among threads implies more work.

As you stated the last thing is to call cudaFreeHost on next epoch change (setEpoch) if on previous epoch the light cache was allocated on host. Maybe it helps to have lightOnHost and light (the address) as class member fields.

For the rest I think is good.